// @shopify/app/dist/cli/services/dev/processes/previewable-extension.js
import { devUIExtensions } from '../extension.js';
import { MANIFEST_VERSION, buildCartURLIfNeeded } from '../../dev.js';
export const launchPreviewableExtensionProcess = async ({ stderr, stdout, abortSignal }, { apiKey, storeFqdn, subscriptionProductUrl, port, cartUrl, proxyUrl, appName, appDotEnvFile, appId, grantedScopes, previewableExtensions, appDirectory, }) => {
    await devUIExtensions({
        appName,
        appDotEnvFile,
        appDirectory,
        id: appId,
        extensions: previewableExtensions,
        stdout,
        stderr,
        signal: abortSignal,
        url: proxyUrl,
        port,
        storeFqdn,
        apiKey,
        grantedScopes,
        checkoutCartUrl: cartUrl,
        subscriptionProductUrl,
        manifestVersion: MANIFEST_VERSION,
    });
};
export async function setupPreviewableExtensionsProcess({ allExtensions, storeFqdn, checkoutCartUrl, ...options }) {
    const previewableExtensions = allExtensions.filter((ext) => ext.isPreviewable);
    if (previewableExtensions.length === 0) {
        return;
    }
    const cartUrl = await buildCartURLIfNeeded(previewableExtensions, storeFqdn, checkoutCartUrl);
    return {
        prefix: 'extensions',
        type: 'previewable-extension',
        function: launchPreviewableExtensionProcess,
        options: {
            pathPrefix: '/extensions',
            port: -1,
            storeFqdn,
            previewableExtensions,
            cartUrl,
            ...options,
        },
    };
}
