// @shopify/app/dist/cli/services/dev/processes/setup-dev-processes.js
import { setupPreviewThemeAppExtensionsProcess } from './theme-app-extension.js';
import { setupPreviewableExtensionsProcess } from './previewable-extension.js';
import { setupDraftableExtensionsProcess } from './draftable-extension.js';
import { setupSendUninstallWebhookProcess } from './uninstall-webhook.js';
import { setupWebProcesses } from './web.js';
import { getAppScopes } from '../../../models/app/app.js';
import { getProxyingWebServer } from '../../../utilities/app/http-reverse-proxy.js';
import { buildAppURLForWeb } from '../../../utilities/app/app-url.js';
import { getAvailableTCPPort } from '@shopify/cli-kit/node/tcp';
export async function setupDevProcesses({ localApp, remoteAppUpdated, token, remoteApp, storeFqdn, commandOptions, network, usesUnifiedDeployment, }) {
    const apiKey = remoteApp.apiKey;
    const apiSecret = remoteApp.apiSecret ?? '';
    const processes = [
        ...(await setupWebProcesses({
            webs: localApp.webs,
            proxyUrl: network.proxyUrl,
            frontendPort: network.frontendPort,
            backendPort: network.backendPort,
            apiKey,
            apiSecret,
            scopes: getAppScopes(localApp.configuration),
        })),
        await setupPreviewableExtensionsProcess({
            allExtensions: localApp.allExtensions,
            storeFqdn,
            apiKey,
            subscriptionProductUrl: commandOptions.subscriptionProductUrl,
            checkoutCartUrl: commandOptions.checkoutCartUrl,
            proxyUrl: network.proxyUrl,
            appName: localApp.name,
            appDotEnvFile: localApp.dotenv,
            grantedScopes: remoteApp.grantedScopes,
            appId: remoteApp.id,
            appDirectory: localApp.directory,
        }),
        await setupDraftableExtensionsProcess({
            localApp,
            remoteApp,
            apiKey,
            token,
            proxyUrl: network.proxyUrl,
            unifiedDeployment: usesUnifiedDeployment,
        }),
        await setupPreviewThemeAppExtensionsProcess({
            allExtensions: localApp.allExtensions,
            storeFqdn,
            apiKey,
            token,
            theme: commandOptions.theme,
            themeExtensionPort: commandOptions.themeExtensionPort,
            notify: commandOptions.notify,
            usesUnifiedDeployment,
        }),
        setupSendUninstallWebhookProcess({
            webs: localApp.webs,
            backendPort: network.backendPort,
            frontendPort: network.frontendPort,
            token,
            storeFqdn,
            apiSecret,
            remoteAppUpdated,
        }),
    ].filter(stripUndefineds);
    // Add http server proxy & configure ports, for processes that need it
    const processesWithProxy = await setPortsAndAddProxyProcess(processes, network.proxyPort);
    // Decide on the appropriate preview URL for a session with these processes
    const anyPreviewableExtensions = processesWithProxy.filter((process) => process.type === 'previewable-extension');
    const previewUrl = anyPreviewableExtensions.length > 0
        ? `${network.proxyUrl}/extensions/dev-console`
        : buildAppURLForWeb(storeFqdn, apiKey);
    return { processes: processesWithProxy, previewUrl };
}
const stripUndefineds = (process) => {
    return process !== undefined && process !== false;
};
async function setPortsAndAddProxyProcess(processes, proxyPort) {
    // Convert processes that use proxying to have a port number and register their mapping rules
    const processesAndRules = await Promise.all(processes.map(async (process) => {
        const rules = {};
        if (process.type === 'web') {
            const targetPort = process.options.portFromConfig || process.options.port;
            rules.default = `http://localhost:${targetPort}`;
            const hmrServer = process.options.hmrServerOptions;
            if (hmrServer) {
                rules.websocket = `http://localhost:${hmrServer.port}`;
                hmrServer.httpPaths.forEach((path) => (rules[path] = `http://localhost:${hmrServer.port}`));
            }
            process.options.port = targetPort;
        }
        else if (process.type === 'previewable-extension') {
            const targetPort = await getAvailableTCPPort();
            rules[process.options.pathPrefix] = `http://localhost:${targetPort}`;
            process.options.port = targetPort;
        }
        return { process, rules };
    }));
    const newProcesses = processesAndRules.map(({ process }) => process);
    const allRules = processesAndRules.map(({ rules }) => rules).reduce((acc, rules) => ({ ...acc, ...rules }), {});
    if (Object.keys(allRules).length > 0) {
        newProcesses.push({
            type: 'proxy-server',
            prefix: 'proxy',
            function: startProxyServer,
            options: {
                port: proxyPort,
                rules: allRules,
            },
        });
    }
    return newProcesses;
}
export const startProxyServer = async ({ abortSignal }, { port, rules }) => {
    const { server } = await getProxyingWebServer(rules, abortSignal);
    await server.listen(port);
};
